robot_3d/src/joint_states_subscriber_3d.cpp

```diff
@@ -0,0 +1,111 @@
+#include <ros/ros.h>
+#include <sensor_msgs/JointState.h>
+#include "std_msgs/String.h"
+
+#include <tf/transform_broadcaster.h>
+
+#include <kdl_parser/kdl_parser.hpp>
+
+#include <kdl/chain.hpp>
+#include <kdl/chainfksolver.hpp>
+#include <kdl/chainfksolverpos_recursive.hpp>
+#include <kdl/frames_io.hpp>
+
+#include <tf_conversions/tf_kdl.h>
+
+class JointMeasurementData
+{
+public:
+  JointMeasurementData(ros::NodeHandle* n)
+  {
+    sub_ = n->subscribe("joint_states", 1000, &JointMeasurementData::jointStatesCallback, this);
+  }
+
+private:
+  void jointStatesCallback(const sensor_msgs::JointState::ConstPtr& msg);
+  // Method to calculate the end effector position using kdl and broadcast the transform.
+  void kdlCalc(double joint_positions[]);
+
+  ros::Subscriber sub_;
+};
+
+void JointMeasurementData::jointStatesCallback(const sensor_msgs::JointState::ConstPtr& msg)
+{
+  ROS_INFO("I heard: [%F]", msg->position[0]);
+  // Array to store the six joint positions from the publisher node.
+  double position[6];
+  position[0] = msg->position[0];
+  position[1] = msg->position[1];
+  position[2] = msg->position[2];
+  position[3] = msg->position[3];
+  position[4] = msg->position[4];
+  position[5] = msg->position[5];
+  // Passing joint positons to method to calculate end effector position.
+  kdlCalc(position);
+}
+
+void JointMeasurementData::kdlCalc(double joint_positions[])
+{
+  // Constructing kdl tree from robot urdf.
+  KDL::Tree fanuc_tree;
```

Comment:
Correct me if I'm wrong (because at this point you know more about kdl than any of us) but it feels like the kdl tree and chain and possibly other things only need to be constructed once at startup and then he exact same tree/chain/etc is used over and over again.

If this is the case then you should not reconstruct the tree each time you are calcutating a new EE Pose. Only create the tree in the class ctor and keep the tree/chain/fksolver object/transform broadcaster as private attributes.

Its worthwhile to think about design in this way: what persists between operations should be setup once and maintained as attributes, and functions will use and possilbly mutate these attributes based on some new and different incoming input.

Reply:
yeah, it really only needs to be done once. I'll make the change :)